### PR TITLE
{Core} Honor scopes specified by Track 2 SDK

### LINKED ISSUE #15179: {Core} AdalAuthentication doesn't honor `scopes` for Track 2 SDK
**Describe the bug**

In Track 2 SDK, `scopes` (`resource`) is managed by SDK, instead of Azure CLI. For example, in [`AzureAppConfigurationClient`](https://github.com/Azure/azure-sdk-for-python/blob/022bc80f2bc645d8d82d647620376fcea2229522/sdk/appconfiguration/azure-appconfiguration/azure/appconfiguration/_azure_appconfiguration_client.py#L128):

```py
if aad_mode:
    scope = base_url.strip("/") + "/.default"
```

But in `get_token`, `scopes` is not honored (discarded), resulting in getting a token for a wrong `scopes` (ARM `https://management.core.windows.net/` by default):

https://github.com/Azure/azure-cli/blob/c2b9a2c5cfe34b108b52b49ac50c6802ac4d729f/src/azure-cli-core/azure/cli/core/adal_authentication.py#L61-L67

### PR DESCRIPTION
**Pull request by jiasli**
Fix #15179

## Issue

In 
- Track 1 SDK, `scopes` (`resource`) is managed by Azure CLI
- Track 2 SDK, `scopes` (`resource`) is managed by SDK

For example, for Track 2 SDK mgmt-plane [`SubscriptionClientConfiguration`](https://github.com/Azure/azure-cli-extensions/blob/120196e953f7b776fa209f4c49838b5cc7af56b0/src/account/azext_account/vendored_sdks/subscription/_configuration.py#L44):

```py
self.credential_scopes = ['https://management.azure.com/.default']
```

For Track 2 SDK data-plane [`AzureAppConfigurationClient`](https://github.com/Azure/azure-sdk-for-python/blob/022bc80f2bc645d8d82d647620376fcea2229522/sdk/appconfiguration/azure-appconfiguration/azure/appconfiguration/_azure_appconfiguration_client.py#L128):

```py
if aad_mode:
    scope = base_url.strip("/") + "/.default"
```

But in `AdalAuthentication.get_token`, `scopes` is not honored (discarded), resulting in getting a token for a wrong `scopes` (ARM `https://management.core.windows.net/` by default):

https://github.com/Azure/azure-cli/blob/c2b9a2c5cfe34b108b52b49ac50c6802ac4d729f/src/azure-cli-core/azure/cli/core/adal_authentication.py#L61-L67

## Fix

This PR fixes the issue by honoring the `scopes` specified by Track 2 SDK.

### User Identity

```sh
# Track 1 SDK mgmt-plan
> az group show -n fy --debug
DEBUG: AdalAuthentication.signed_session invoked by Track 1 SDK

# Track 2 SDK mgmt-plane
> az account subscription show --id 0b1f6471-1bf0-4dda-aec3-cb9272f09590 --debug
DEBUG: AdalAuthentication.get_token invoked by Track 2 SDK with scopes=('https://management.azure.com/.default',)

# Track 1 SDK data-plane
> az account subscription show --id 0b1f6471-1bf0-4dda-aec3-cb9272f09590 --debug
DEBUG: Profile.get_raw_token invoked with resource='https://storage.azure.com', subscription=None, tenant=None

# Track 2 SDK data-plane
az storage blob list --account-name st0921 -c con1 --auth-mode login --debug
DEBUG: AdalAuthentication.get_token invoked by Track 2 SDK with scopes=('https://storage.azure.com/.default',)
```

### Managed Identity

```sh
# Track 1 SDK mgmt-plan
> az group show -n fy --debug
DEBUG: MSIAuthenticationWrapper.signed_session invoked by Track 1 SDK

# Track 2 SDK mgmt-plane
> az account subscription show --id 0b1f6471-1bf0-4dda-aec3-cb9272f09590 --debug
DEBUG: MSIAuthenticationWrapper.get_token invoked by Track 2 SDK with scopes=('https://management.azure.com/.default',)

# Track 1 SDK data-plane
> az account subscription show --id 0b1f6471-1bf0-4dda-aec3-cb9272f09590 --debug
DEBUG: Profile.get_raw_token invoked with resource='https://storage.azure.com', subscription=None, tenant=None

# Track 2 SDK data-plane
az storage blob list --account-name st0921 -c con1 --auth-mode login --debug
DEBUG: MSIAuthenticationWrapper.get_token invoked by Track 2 SDK with scopes=('https://storage.azure.com/.default',)
```

### CODE REVIEW

**Comment by yonzhan** [comment]
Core

File: src/azure-cli/azure/cli/command_modules/storage/_validators.py [hunk 1]
```diff
@@ -129,8 +129,7 @@ def validate_client_parameters(cmd, namespace):
             if is_storagev2(prefix):
                 from azure.cli.core._profile import Profile
                 profile = Profile(cli_ctx=cmd.cli_ctx)
-                n.token_credential, _, _ = profile.get_login_credentials(
-                    resource="https://storage.azure.com", subscription_id=n._subscription)
+                n.token_credential, _, _ = profile.get_login_credentials(subscription_id=n._subscription)
```
Comment (jiasli) [inline src/azure-cli/azure/cli/command_modules/storage/_validators.py:132]: `scope` is now managed by [Track 2 Storage SDK](https://github.com/Azure/azure-sdk-for-python/blob/900aa7d23212f528aec352b9bff2a87cd5294720/sdk/storage/azure-storage-blob/azure/storage/blob/_shared/constants.py#L25) (azure/multiapi/storagev2/blob/v2019_12_12/_shared/constants.py:25):

```py
STORAGE_OAUTH_SCOPE = "https://storage.azure.com/.default"
```

**Comment by jiasli** [comment]
Unfortunately, due an incorrect `credential_scopes` managing logic in existing SDKs (https://github.com/Azure/azure-sdk-for-python/issues/12947), even though the bug has been fixed in https://github.com/Azure/autorest.python/pull/745, merging this PR will break all commands using **Track 2 SDKs with sovereign clouds**, as `credential_scopes` will be passed to `scopes` as 

```
['https://management.azure.com/.default', 'https://management.core.chinacloudapi.cn/.default']
```

`'https://management.core.chinacloudapi.cn/.default'` will be discarded.

File: src/azure-cli-core/azure/cli/core/adal_authentication.py [hunk 2]
```diff
@@ -60,14 +66,24 @@ def _get_token(self):
...
     def get_token(self, *scopes, **kwargs):  # pylint:disable=unused-argument
-        _, token, full_token, _ = self._get_token()
+        logger.debug("AdalAuthentication.get_token invoked by Track 2 SDK with scopes=%s", scopes)
+
+        # Deal with an old Track 2 SDK issue where the default credential_scopes is extended with
+        # custom credential_scopes. Instead, credential_scopes should be replaced by custom credential_scopes.
+        # https://github.com/Azure/azure-sdk-for-python/issues/12947
+        # We simply remove the first one if there are multiple scopes provided.
+        if len(scopes) > 1:
+            scopes = scopes[1:]
```
Comment (jiasli) [inline src/azure-cli-core/azure/cli/core/adal_authentication.py:76]: Add a patch to handle issue https://github.com/Azure/azure-sdk-for-python/issues/12947 in old Track 2 SDKs.

**Review by zhoxing-ms — APPROVED** [review]

**Review by evelyn-ys — APPROVED** [review]

File: src/azure-cli-core/azure/cli/core/util.py [hunk 3]
```diff
@@ -1165,3 +1165,36 @@ def handle_version_update():
...
+def scopes_to_resource(scopes):
+    """Convert MSAL scopes to ADAL resource by stripping the /.default suffix and return a str.
+    For example: ['https://management.core.windows.net/.default'] -> 'https://management.core.windows.net'
+
+    :param scopes: The MSAL scopes. It can be a list or tuple of string
+    :return: The ADAL resource
+    :rtype: str
+    """
+    scope = scopes[0]
+    if scope.endswith(".default"):
```
Comment (arrownj) [inline src/azure-cli-core/azure/cli/core/util.py:1195]: Actually we have some cases that scope is not `.default`. This function will not cover those cases ?

**Review by bim-msft — APPROVED** [review]

File: src/azure-cli-core/azure/cli/core/util.py (hunk 3 above)
Comment (jiasli) [reply src/azure-cli-core/azure/cli/core/util.py:1195]: Could you give a pointer to "some cases"?

Comment (jiasli) [reply src/azure-cli-core/azure/cli/core/util.py:1195]: Beside, if `scope` doesn't end with `.default`, it can't be converted to `resource` either.

**Review by arrownj — APPROVED** [review]

File: src/azure-cli-core/azure/cli/core/util.py (hunk 3 above)
Comment (jsntcy) [inline src/azure-cli-core/azure/cli/core/util.py:1195]: >if scope.endswith(".default"): [](start = 4, length = 30)

Do we make this case insensitive?

File: src/azure-cli-core/azure/cli/core/util.py [hunk 4]
```diff
@@ -1165,3 +1165,36 @@ def handle_version_update():
...
+        scope = scope[:-len(".default")]
+
+    # Trim extra ending slashes. https://datalake.azure.net// -> https://datalake.azure.net/
+    scope = scope.rstrip('/') + '/'
+    return scope
```
Comment (jsntcy) [inline src/azure-cli-core/azure/cli/core/util.py:1200]: Is it expected that "scope" is [https://management.core.windows.net/](https://management.core.windows.net/) or it should be [https://management.core.windows.net](https://management.core.windows.net)?

File: src/azure-cli-core/azure/cli/core/util.py (hunk 3 above)
Comment (jiasli) [reply src/azure-cli-core/azure/cli/core/util.py:1195]: No. `.default` is the only supported form. See [The /.default scope](https://docs.microsoft.com/en-us/azure/active-directory/develop/v2-permissions-and-consent#the-default-scope)

File: src/azure-cli-core/azure/cli/core/util.py [hunk 5]
```diff
@@ -1165,3 +1165,36 @@ def handle_version_update():
...
+
+def scopes_to_resource(scopes):
+    """Convert MSAL scopes to ADAL resource by stripping the /.default suffix and return a str.
+    For example: ['https://management.core.windows.net/.default'] -> 'https://management.core.windows.net'
+
+    :param scopes: The MSAL scopes. It can be a list or tuple of string
+    :return: The ADAL resource
+    :rtype: str
+    """
+    scope = scopes[0]
```
Comment (jsntcy) [inline src/azure-cli-core/azure/cli/core/util.py:1194]: Do we only support one scope now as it is a list?

Comment (jiasli) [reply src/azure-cli-core/azure/cli/core/util.py:1194]: Per [Request an access token](https://docs.microsoft.com/en-us/azure/active-directory/develop/v2-oauth2-auth-code-flow#request-an-access-token),

> The scopes must all be from a single resource.

File: src/azure-cli-core/azure/cli/core/util.py (hunk 4 above)
Comment (jiasli) [reply src/azure-cli-core/azure/cli/core/util.py:1200]: Yes. See https://github.com/Azure/azure-cli/blob/4e1ff0ec626ea46d74793ad92a1b5eddc2b6e45b/src/azure-cli-core/azure/cli/core/cloud.py#L303

File: src/azure-cli-core/azure/cli/core/util.py [hunk 6]
```diff
@@ -1165,3 +1165,36 @@ def handle_version_update():
...
         logger.warning(ex)
+
+
+def resource_to_scopes(resource):
+    """Convert the ADAL resource ID to MSAL scopes by appending the /.default suffix and return a list.
+    For example: 'https://management.core.windows.net/' -> ['https://management.core.windows.net/.default']
+    :param resource: The ADAL resource ID
+    :return: A list of scopes
+    """
+    if 'datalake' in resource or 'batch' in resource or 'database' in resource:
```
Comment (jsntcy) [inline src/azure-cli-core/azure/cli/core/util.py:1176]: Is it possible that we move this tricky logic from core to module level?

File: src/azure-cli-core/azure/cli/core/util.py [hunk 7]
```diff
@@ -1165,3 +1165,36 @@ def handle_version_update():
...
+        # TODO: This should be fixed on the service side.
+        scope = resource + '/.default'
+    else:
+        scope = resource.rstrip('/') + '/.default'
+    return [scope]
+
+
+def scopes_to_resource(scopes):
+    """Convert MSAL scopes to ADAL resource by stripping the /.default suffix and return a str.
+    For example: ['https://management.core.windows.net/.default'] -> 'https://management.core.windows.net'
```
Comment (jsntcy) [inline src/azure-cli-core/azure/cli/core/util.py:1188]: should be [https://management.core.windows.net/](https://management.core.windows.net/)

Comment (jiasli) [reply src/azure-cli-core/azure/cli/core/util.py:1188]:
```suggestion
    For example: ['https://management.core.windows.net/.default'] -> 'https://management.core.windows.net/'
```

**Review by jsntcy — APPROVED** [review]
:shipit:

File: src/azure-cli-core/azure/cli/core/util.py (hunk 6 above)
Comment (jiasli) [reply src/azure-cli-core/azure/cli/core/util.py:1176]: Good point. Currently `resource_to_scopes` is only called by `_get_mgmt_service_client`. We should actually remove this tricky logic totally from this function. Will do in a separate PR.

**Pull request merged**